Spec Balancer hard fork

dapplion opened the pull request:
Spec Balancer fork, and the specific change with a reference document in the style of an EIP

dapplion committed:
Spec Balancer hard fork

filoozom commented on execution/balancer_recovery.md:
Is this `if` condition necessary? The result would be the same with or without, even if they are already identical (which shouldn't happen either way). I guess that this also assumes that the code hash gets updated automatically by setting the bytecode?

dapplion replied:
Right, I can exclude

filoozom commented on execution/balancer_recovery.md:
A nit, but this isn't really bytecode. Maybe call it "Solidity contract" and then mention that the bytecode is generated from this contract?

dapplion replied:
Sure, can you share the exact solc compilers flags you used to generate the final bytecode?

filoozom commented on execution/balancer_recovery.md:
```suggestion
- The attacker can no longer issue valid transactions from the compromised EOA (enforced by [EIP-3607](https://eips.ethereum.org/EIPS/eip-3607), and
```

filoozom commented on execution/balancer_recovery.md:
```suggestion
Thanks to [EIP-3607](https://eips.ethereum.org/EIPS/eip-3607), even with the attacker controlling the private key of `BALANCER_ATTACKER_ADDRESS` any transaction from it can't be included in blocks after the hard-fork.
```

filoozom commented on execution/balancer_recovery.md:
```suggestion
- It has a single immutable owner (`BALANCER_RESCUE_ADDRESS`).
```

filoozom commented on execution/balancer_recovery.md:
```suggestion
- It forwards value and calldata.
- It doesn't contain any state.
```

filoozom commented on execution/balancer_recovery.md:
```suggestion
The bytecode and its Solidity source should be independently verified by multiple client teams and external auditors.
```

filoozom commented on execution/balancer_recovery.md:
```suggestion
| `BALANCER_HARDFORK_TIMESTAMP` | `1766419900` |
```

filoozom reviewed (changes requested):
Suggested a few changes. Some EIPs has brackets but no link, and some EIPs don't have brackets. I added links to the Ethereum specs, but we should probably be consistent in having or not having links for each EIP reference.

filoozom commented on network-upgrades/balancer.md:
```diff
@@ -0,0 +1,35 @@
+
+## Included changes
+
+This hard fork introduces an irregular state change to recover the stolen funds of the Balancer V2 hack on November 2025.
```
```suggestion
This hard fork specification introduces an irregular state change intended to recover funds stolen in the Balancer V2 hack in November 2025. Whether and when this specification is activated on Gnosis Chain depends on validator adoption and client configuration.
```

dapplion committed:
Apply suggestions from code review

Co-authored-by: Philippe Schommers <philippe@schommers.be>

dapplion committed:
Change 'Bytecode' section to 'Solidity contract'

Updated documentation to reflect Solidity contract details.

dapplion committed:
Simplify attacker account bytecode assignment

dapplion committed:
Enhance balancer_recovery.md with contract details

Added information about the test version of the contract and its deployment address.

dapplion committed:
Clarify Solidity contract details and deployment info

Updated the Solidity contract section with specific compiler version and optimizations details. Added information about the deployed test version of the contract and reproducibility instructions.

filoozom commented on execution/balancer_recovery.md:
```suggestion
`BALANCER_RESCUE_BYTECODE` is generated by compiling the following Solidity source file with `solc` version `v0.8.30+commit.73712a01` and no optimizations with 200 runs.
```

filoozom reviewed (dismissed):
Probably just a nit, but all else looks good to me now!

dapplion committed:
Update execution/balancer_recovery.md

Co-authored-by: Philippe Schommers <philippe@schommers.be>
